Add Kubevirt provider

nirarg opened the pull request:
This PR is part of task done to add KubeVirt platform as infra provider for Openshift.

The change introduced in this PR purpose is to provide a way to install Openshift on KubeVirt infrastructure

For more information, please find the enhancement PR:
openshift/enhancements#417

Related PRs/changes done as part of this task:

1. Created openshift/cluster-api-provider-kubevirt repository, hosts an implementation of a provider for KubeVirt for the OpenShift machine-api: https://github.com/openshift/cluster-api-provider-kubevirt
Also adding this project's image to openshift is in progress
2. openshift/installer: https://github.com/openshift/installer/pull/4350
3. openshift/cloud-credential-operator: openshift/cloud-credential-operator#260
4. openshift/machine-config-operator: openshift/machine-config-operator#2098
5. openshift/api: openshift/api#734

Additional job is done to have tests and CI coverage (work in progress in github.com/openshift/release repository)

rgolangh commented:
images creation depends on https://github.com/openshift/release/pull/12318 (hopefully merged soon)

chenyosef commented on install/0000_30_machine-api-operator_00_credentials-request.yaml:
```diff
@@ -151,3 +151,19 @@ spec:
+spec:
+  secretRef:
+    namespace: openshift-machine-api
+    # This is the of the new secret
```
Should be "This is the **name** of the new secret"

nirarg replied:
I removed the comment

nirarg commented:
/retest

JoelSpeed commented:
Changes here look fine, please add a description to the PR before we merge

We have a script that pushes updates to MAO out to repositories that depend on this repo, you may want to add kubevirt to this list
https://github.com/openshift/machine-api-operator/blob/1842a5506d152f88dff27e536bec461d21eeec4f/hack/push-updates.sh#L45-L51

nirarg commented:
Hi @JoelSpeed, thank you for the review
I added the description and also added kubevirt to the push-updates script
I will leave the [wip] flag until `cluster-api-provider-kubevirt` image would be added to openshift release

JoelSpeed commented:
@nirarg Great, please ping me once that's done and you're ready to look at getting this merged, I don't think I have any further feedback at this time

nirarg commented:
/retest

nirarg commented:
/retest

nirarg commented:
Hi @JoelSpeed , I removed the `WIP` tag, can you please review this again?

JoelSpeed commented:
/approve

One question I do have though, seems we are using `Kubevirt` as the camel case capitalisation, is that correct? I thought it was `KubeVirt` with a capital V for the product name? Is `kubeVirt` not allowed? (Asking because we use `VSphere` and `vSphere` within the codebase because of the product name)

nirarg commented:
Hi @JoelSpeed 
Thank you for the review/approve
I use `KubeVirt` when it is standalone parameter (defined in openshift/api repository)
I used `Kubevirt` when it part of longer parameter name (like `KubevirtPlatformType`), which there I followed the camel case capitalisation.

chenyosef commented:
/lgtm
